core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala

```diff
@@ -34,18 +37,31 @@ private[spark] class HistoryPage(parent: HistoryServer) extends WebUIPage("") {
     val requestedIncomplete =
       Option(request.getParameter("showIncomplete")).getOrElse("false").toBoolean
 
-    val allApps = parent.getApplicationList().filter(_.completed != requestedIncomplete)
-    val actualFirst = if (requestedFirst < allApps.size) requestedFirst else 0
-    val apps = allApps.slice(actualFirst, Math.min(actualFirst + pageSize, allApps.size))
-
+    val allCompletedAppsNAttempts = 
+        parent.getApplicationList().filter(_.completed != requestedIncomplete)
+    val (hasAttemptInfo, appToAttemptMap)  = getApplicationLevelList(allCompletedAppsNAttempts)
```

Comment:
Hi,

If at a particular time, there are few event logs, which have attempt info, and there are few, which do no have. Then we display the same number of columns irrespective of the event log having attempt info or not.

To do above, we need just one flag of hasAttemptInfo for all the applications we have, and it is not per application. If we do that map logic, then we will need to iterate again on that map, after creating the same.

Regarding that memory exploding, if we first slice, and at a time ( in the case, when history server is restarting with some event logs already there ), then we might see one page's UI without attempt info, and another page with attempt UI. Is it fine?